hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimelineUtils.java

```diff
@@ -295,4 +296,24 @@ public static Option<HoodieInstant> getEarliestInstantForMetadataArchival(
       return Option.empty();
     }
   }
+
+  /**
+   * Find the earliest incomplete commit, deltacommit, or non-clustering replacecommit,
+   * so that the incremental pulls should be strictly before this instant.
+   * This is to guard around multi-writer scenarios where a commit starting later than
+   * another commit from a concurrent writer can finish earlier, leaving an inflight commit
+   * before a completed commit.
+   */
+  public static HoodieTimeline filterTimelineForIncrementalQueryIfNeeded(
+      HoodieTableMetaClient metaClient, HoodieTimeline completedCommitTimeline) {
+    Option<HoodieInstant> firstIncompleteCommit = metaClient.getCommitsTimeline()
+        .filterInflightsAndRequested()
+        .filter(instant ->
```

Comment:
Could this firstly find the earliest incomplete commit, deltacommit? If there is no the earliest incomplete commit, deltacommit, then find the earliest non-clustering replacecommit. Because there is certain performance consuming for `ClusteringUtils.getClusteringPlan(metaClient, instant)`.

Reply:
non-clustering replacecommit could be the earliest so we need to check all cases together. the perf shouldn't be a concern as the number of total inflight/requested instants is very limited in practice. besides it'll be short-cutted by the first condition check